text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+    ```
+    e = Blake2B(proof_tag_e | commitment | d)
+    ```
+
+3. The prover computes the following, and includes these values in the Proof of Opening:
+
+    ```
+    u = y + e * token_id
+    v = s + e * blinding_factor
+    ```
+
+4. The verifier of the Proof of Opening must verify:
+
+    ```
+    G * u + H * v = d + commitment * e
+    ```
+
+    - This property follows a similar property of the [Schnorr Proof of Knowledge protocol](https://en.wikipedia.org/wiki/Proof_of_knowledge#Schnorr_protocol)
+
+    
+This protocol is non-interactive because the prover anticipates the verifier's `e` value, so requires no interaction from the verifier.
+
+Thus, the Proof of Knowledge could be represented, for example, with:
+
+```
+struct ProofOfKnowledge {
+
+    /// Prover value established from hashes of commitment and blinding factor secrets
+    CompressedRistretto d,
+    
+    /// Prover value from the token_id response to the verifier's commitment-derived challenge
+    CurveScalar u,
+    
+    /// Prover value from the blinding_factor response to the verifier's commitment-derived challenge
+    CurveScalar v,
+}
+```
+
+- Implementation note: The proof of knowledge should be constructed at [this line](https://github.com/mobilecoinfoundation/mobilecoin/blob/a9db0bc7bcac9bac1c1232b194003d02bccaf283/transaction/core/src/ring_signature/rct_bulletproofs.rs#L304) in the sign function.
+
+Then, the Transaction Validator need perform the following validation check, after calculating the sum of the outputs (in order to verify the sum of the inputs matches the sum of the outputs, and thus that no value was created or destroyed), at [this line](https://github.com/mobilecoinfoundation/mobilecoin/blob/a9db0bc7bcac9bac1c1232b194003d02bccaf283/transaction/core/src/ring_signature/rct_bulletproofs.rs#L170):
+
+```
+verify(G, H, commitment, proof_of_knowledge) {
+    e = Blake2B(proof_tag_e | commitment | proof_of_knowledge.d)
+    proof_of_knowledge.d + (commitment * e) == G * proof_of_knowledge.u + H * proof_of_knowledge.v
+}
+```
+
+### Fee Outputs
+
+Currently, the enclave aggregates the fees for multiple transactions in a block, in order to mint a single fee output for the block and conserve space. To support multiple confidential TokenTypes, when minting fee outputs, the enclave must create multiple fee outputs if multiple TokenTypes were used in the block. A consideration here is that we would like for an observer not to discern whether a block contains transactions of multiple asset types by statistically analyzing the number of outputs to derive information about the number of fee outputs. 
+
+A simple proposal is that the enclave discontinue fee aggregation, so that the number of fee outputs scale linearly with the number of transactions in a block (not output TXOs). For example, a block with 3 transactions includes 3 fee outputs. Because each transaction can mint up to 16 output TXOs, there may be a variable number of outputs in the block. 
```

Comment:
This would amplify the problem I discussed in the [fee epochs MCIP](https://github.com/mobilecoinfoundation/mcips/pull/10).